sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,833 @@
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2017 Esteban Tovagliari, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+#define APPLESEED_ENABLE_IMATH_INTEROP
+
+#include "renderer/global/globallogger.h"
+
+// appleseed.renderer headers.
+//#include "renderer/api/bsdf.h"
+#include "renderer/api/object.h"
+//#include "renderer/api/material.h"
+#include "renderer/api/project.h"
+#include "renderer/api/scene.h"
+#include "renderer/api/utility.h"
+//#include "renderer/api/log.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/matrix.h"
+#include "foundation/math/triangulator.h"
+#include "foundation/math/transform.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/autoreleaseptr.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/string.h"
+
+// appleseed.main headers.
+#include "main/dllvisibility.h"
+
+// AlembicAssembly headers.
+#include "alembicassembly.h"
+
+#include <Alembic/Abc/All.h>
+#include <Alembic/AbcCoreFactory/All.h>
+#include <Alembic/AbcGeom/All.h>
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+#include <vector>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+namespace
+{
+
+// roll into given matrix stack and return the flatten one
+asf::Matrix4d flatten_xform(std::vector<asf::Matrix4d> mtx_stack)
+{
+    // the matrix we will return
+    auto out_mtx = asf::Matrix4d::identity();
+
+    // mutiply the each matrix of the stack from bottom to top
+    for(auto mtx : mtx_stack)
+    {
+        out_mtx = mtx * out_mtx;
+    }
+
+    return out_mtx;
+}
+
+// roll into given xform sequence stack and return the flatten one
+asr::TransformSequence flatten_xform_seq(const std::vector<asr::TransformSequence>& xform_seq_stack)
+{
+    auto out_xform_seq = asr::TransformSequence();
```

Comment:
I think this creates a copy.

Simply do this:

    asr::TransformSequence out_xform_seq;

Reply:
I suggest moving the transform stack into its own class.
You can optimize the code if instead of pushing a transform and then looping to flatten it everytime is needed,
you push matrix * stack.top(). Then you don't need to loop to flatten.
You can also simplify it by considering a simple matrix as a transform sequence with only one sample and unify the two cases.